MaintenanceJobDesignDoc.md

```diff
@@ -0,0 +1,247 @@
+for the first time in 31 days, then that large prefetch pack is referenced and
+all older prefetch packs are "stale" and not referenced by the multi-pack-index.
+
+We can extend the `git multi-pack-index` builtin to delete pack-files that have
+no objects referenced by the multi-pack-index. This should happen at the same
+time as we update the multi-pack-index since we don't want to have this list of
+pack-files in the multi-pack-index if they don't exist. We also don't want to
+remove them from the multi-pack-index while the pack-files still exist, because
+then a Git process will start referencing those pack-files in the `packed_git`
+linked list.
+
+### Repack the prefetch pack-files
+
+We can reduce the number of prefetch packs and improve file locality by
+combining prefetch packs into new "repacked" pack-files. Use `git show-index
+<object-dir>/pack/prefetch-<timestamp[i]>-<hash[i]>.idx` to list the objects
+in the prefetch packs and pipe that output to `git pack-objects -q
+<object-dir>/pack/repacked-<timestamp>`.
+
+We should refrain from packing the most-recent prefetch packs, as we don't want
+to get in the way of the prefetch machinery for determining the most-recent
+timestamp. We could, however, start repacking the prefetch packs from
+oldest-to-newest, stopping in the most-recent week, or when the prefetch packs
+are large enough to satisfy a batch. Some stopping conditions could include
+"pack-file size is above 1GB".
+
+Scheduling Actions
+------------------
+
+We already have a prefetch action that is automatically queued in the
+background. In addition, a `gvfs prefetch --commits` action sends a message
+to the mount process to run a post-fetch job (compute multi-pack-index and
+commit-graph). These actions already have some logic that we will want to
+replicate, so we should make them explicit requirements of an action:
+
+* Most actions intend to have some delay between successful runs. Use a
+  `IsReady(DateTime time)` method to perform only the check if the given
+  `time` is long enough after the previous run. (This is most likely stored
+  as a timestamp in a list of pack-files or as the content of a file in
+  the shared object directory.)
+* Most logic happens during an `Execute()` method.
+* If the mount process needs to shut down while the action is active, we need
+  a `Stop()` method for the action to interrupt itself. This includes halting
+  any Git processes that are being run.
+
+The actions are then run in sequence by an `ActionRunner` that triggers a run
+even on some interval, runs the actions that are ready, and tracks the
+currently-running action (so the mount process can tell the `ActionRunner` to
+`Stop()` and it will stop the running action). The `ActionRunner` can keep an
+ordered list of actions, and during each run of the background task it can
+check `IsReady(DateTime.UtcNow)` and run `Execute()` as necessary.
+
+Here are some initial ideas for the frequency of each action:

```

Comment:
Something that might be worth considering (though I don't know how difficult it would be) is checking the state of the machine as well.  If we were able to give preference to times when the machine is locked (or the monitor has gone to sleep) we might be able to reduce this chances of these BG tasks taking resources from whatever the user happens to be doing on the machine.

Reply:
Related: Currently can multiple prefetch actions run at the same time (if repos are using different .gvfsCache folders, or different subfolders of the same .gvfsCache folder)?  

If so, I wonder if we should have any cross-repo coordination to avoid having more than one of these running at once (we could also wait and see if anyone actually hits and perf\memory issues from this scenario).

Reply:
@derrickstolee would be good to know the result of your "Is the machine awake" investigation.  As you may be aware we are currently getting a lot of "Device is not ready" exceptions.  Do we expect success when the machine is locked?

Reply:
>where would we put it? Do we have some application-level storage?

Currently our machine-wide data lives in C:\ProgramData\GVFS.  The one tricky part about that folder though is you *might* need elevated permissions to write to it.  I'm not sure if that can be adjusted on a per-file basis.

> For now, I think we should keep this global lock in the back of our minds, and instead focus on making the jobs run infrequently and in reasonably-sized batches.

Sounds good

Reply:
Good point! If every "idle" time actually locks the drive, then we will be sunk.